File: Java/src/Protocol.java
public abstract class Protocol {
    StringBuffer type = new StringBuffer(1);
    StringBuffer sensorName = new StringBuffer(8);
    private StringBuffer checksum = new StringBuffer(4);

    void setType(String type) {
        this.type = new StringBuffer(type);
    }

    String getType() {
        return type.toString();
    }

    void setSensorName(String sensorName) {
        this.sensorName = new StringBuffer(sensorName);
    }

    String getSensorName() {
        return sensorName.toString();
    }

    void setChecksum(String checksum) {
        this.checksum = new StringBuffer(checksum);
    }

    int getChecksum() {
        return Integer.valueOf(checksum.toString());
    }

    protected abstract Boolean checksumIsValid();

    int sumChunk(StringBuffer chunk) {
        int total = 0;

        char [] _chunk = chunk.toString().toCharArray();

        for (char letter: _chunk) {
            total += (int)letter;
        }

        return total;
    }

    abstract String getLog();
}

File: cpp/protocol.cc
#include "protocol.hpp"
#include "updateprotocol.hpp"
#include "requestprotocol.hpp"

static inline bool isProtocolValid(Protocol* protocol) ;

Protocol* Protocol::getProtocol(string message) {
    Protocol *protocol;

    if ((message.length() == 35) && (message[0] == 'U')) {
        protocol =  new UpdateProtocol(message);
    } else if ((message.length() == 21) && (message[0] == 'R')) {
        protocol = new RequestProtocol(message);
    } else {
        protocol = nullptr;
    }

    if ((protocol != nullptr) && isProtocolValid(protocol)) {
        return protocol;
    } else {
        return nullptr;
    }
}

static inline bool isProtocolValid(Protocol* protocol) {
    return protocol->checksumIsValid();
}

int Protocol::sumChunk(string chunk) {
    int total = 0;

    for (char letter: chunk) {
        total += (int)letter;
    }

    return total;
}

char Protocol::getType() {
    return type[0];
}

string Protocol::getSensorName() {
    return sensorName;
}